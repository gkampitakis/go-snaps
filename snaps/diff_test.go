package snaps

import (
	"strings"
	"testing"

	"github.com/gkampitakis/go-snaps/internal/colors"
	"github.com/gkampitakis/go-snaps/internal/test"
)

var a = `Proin justo libero, pellentesque sit amet scelerisque ut, sollicitudin non tortor. 
		Sed velit elit, accumsan sed porttitor nec, elementum quis sapien. 
		Phasellus mattis purus in dui pretium, eu euismod metus feugiat. 
		Morbi turpis tellus, tincidunt mollis rutrum at, mattis laoreet lacus. 
		Donec in quam tempus, eleifend erat sit amet, aliquet metus. 
		Sed ullamcorper velit a est efficitur, et tempus ante rhoncus. 
		Aliquam diam sapien, vulputate sit amet elit sit amet, commodo eleifend sapien. 
		Donec consequat at nibh id mattis. Quisque vitae sagittis eros, convallis consectetur ante. 
		Duis finibus suscipit mi sed consectetur. Nulla libero neque, sagittis vel nulla vel,
		 vestibulum sagittis mauris. Ut laoreet urna lectus. 
		 Sed lorem felis, condimentum eget vehicula non, sagittis sit amet diam. 
		 Vivamus ut sapien at erat imperdiet suscipit id a lectus.`

var b = `Proin justo libero, pellentesque sit amet scelerisque ut, sollicitudin non tortor. 
		Sed velit elit, accumsan sed Ipsum nec, elementum quis sapien. 
		Phasellus mattis purus in dui pretium, eu euismod metus feugiat. 
		Morbi turpis Lorem, tincidunt mollis rutrum at, mattis laoreet lacus. 
		Donec in quam tempus, eleifend erat sit amet, aliquet metus. 
		Sed ullamcorper velit a est efficitur, et tempus ante rhoncus. 
		Aliquam diam sapien, vulputate sit amet elit sit amet, commodo eleifend sapien. 
		Donec consequat at nibh id mattis. Quisque vitae sagittis eros, convallis consectetur ante. 
		Duis finibus suscipit mi sed consectetur. Nulla libero neque, sagittis vel nulla vel,
		vestibulum sagittis mauris. Ut laoreet urna lectus. 
		Sed lorem felis, condimentum eget vehicula non, sagittis sit amet diam. 
		Vivamus ut sapien at erat imperdiet suscipit id a lectus.
		Another Line added.`

func TestStringUtils(t *testing.T) {
	t.Run("splitNewlines", func(t *testing.T) {
		for _, v := range []struct {
			input    string
			expected []string
		}{
			{"foo", []string{"foo\n"}},
			{"foo\nbar", []string{"foo\n", "bar\n"}},
			{"foo\nbar\n", []string{"foo\n", "bar\n", "\n"}},
			{`abc
			efg
			hello \n world`, []string{"abc\n", "\t\t\tefg\n", "\t\t\thello \\n world\n"}},
		} {
			v := v
			t.Run(v.input, func(t *testing.T) {
				t.Parallel()
				test.Equal(t, v.expected, splitNewlines(v.input))
			})
		}
	})

	t.Run("isSingleLine", func(t *testing.T) {
		test.Equal(t, true, isSingleline("hello world"))
		test.Equal(t, true, isSingleline("hello world\n"))
		test.Equal(t, false, isSingleline(`hello 
		 world
		 `))
		test.Equal(t, false, isSingleline("hello \n world\n"))
		test.Equal(t, false, isSingleline("hello \n world"))
	})
}

func TestDiff(t *testing.T) {
	t.Run("should return empty string if no diffs", func(t *testing.T) {
		t.Run("single line", func(t *testing.T) {
			expected, received := "Hello World\n", "Hello World\n"

			if diff := prettyDiff(expected, received); diff != "" {
				t.Errorf("found diff between same string %s", diff)
			}
		})

		t.Run("multiline", func(t *testing.T) {
			expected := `one snapshot
			containing new lines
			`
			received := expected

			if diff := prettyDiff(expected, received); diff != "" {
				t.Errorf("found diff between same string %s", diff)
			}
		})
	})

	t.Run("should print header consistently", func(t *testing.T) {
		MatchSnapshot(t, header(10000, 20))
		MatchSnapshot(t, header(20, 10000))
	})

	t.Run("with color", func(t *testing.T) {
		t.Run("should apply highlights on single line diff", func(t *testing.T) {
			a := strings.Repeat("abcd", 20)
			b := strings.Repeat("abcf", 20)

			MatchSnapshot(t, prettyDiff(a, b))
		})

		t.Run("multiline diff", func(t *testing.T) {
			MatchSnapshot(t, prettyDiff(a, b))
		})
	})

	t.Run("no color", func(t *testing.T) {
		t.Cleanup(func() {
			colors.NOCOLOR = false
		})
		colors.NOCOLOR = true

		t.Run("should apply highlights on single line diff", func(t *testing.T) {
			a := strings.Repeat("abcd", 20)
			b := strings.Repeat("abcf", 20)

			MatchSnapshot(t, prettyDiff(a, b))
		})

		t.Run("multiline diff", func(t *testing.T) {
			MatchSnapshot(t, prettyDiff(a, b))
		})
	})
}
