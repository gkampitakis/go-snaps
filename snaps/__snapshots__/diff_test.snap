
[TestDiff/should_build_diff_report_consistently - 1]

[38;5;52m[48;5;225m- Snapshot    - 20[0m
[38;5;22m[48;5;159m+ Received + 10000[0m

mock-diff
[2mat snap/path:10
[0m
---

[TestDiff/should_build_diff_report_consistently - 2]

[38;5;52m[48;5;225m- Snapshot - 10000[0m
[38;5;22m[48;5;159m+ Received    + 20[0m

mock-diff
[2mat snap/path:20
[0m
---

[TestDiff/with_color/should_apply_highlights_on_single_line_diff - 1]

[38;5;52m[48;5;225m- Snapshot - 20[0m
[38;5;22m[48;5;159m+ Received + 20[0m

[48;5;225m[38;5;52m- [0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m[48;5;225m[38;5;52mabc[0m[48;5;127m[38;5;255md[0m
[48;5;159m[38;5;22m+ [0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m[48;5;159m[38;5;22mabc[0m[48;5;23m[38;5;255mf[0m

[2mat snap/path:10
[0m
---

[TestDiff/with_color/multiline_diff - 1]

[38;5;52m[48;5;225m- Snapshot - 5[0m
[38;5;22m[48;5;159m+ Received + 6[0m

[33;1m@@ -1,12 +1,13 @@[0m

  [2mProin justo libero, pellentesque sit amet scelerisque ut, sollicitudin non tortor. 
[0m[48;5;225m[38;5;52m- [0m[48;5;225m[38;5;52m     Sed velit elit, accumsan sed [0m[48;5;127m[38;5;255mporttitor[0m[48;5;225m[38;5;52m nec, elementum quis sapien. [0m
[48;5;159m[38;5;22m+ [0m[48;5;159m[38;5;22m         Sed velit elit, accumsan sed [0m[48;5;23m[38;5;255mIpsum[0m[48;5;159m[38;5;22m nec, elementum quis sapien. [0m
  [2m                                                   Phasellus mattis purus in dui pretium, eu euismod metus feugiat. 
[0m[48;5;225m[38;5;52m- [0m[48;5;225m[38;5;52m     Morbi turpis [0m[48;5;127m[38;5;255mtellus[0m[48;5;225m[38;5;52m, tincidunt mollis rutrum at, mattis laoreet lacus. [0m
[48;5;159m[38;5;22m+ [0m[48;5;159m[38;5;22m         Morbi turpis [0m[48;5;23m[38;5;255mLorem[0m[48;5;159m[38;5;22m, tincidunt mollis rutrum at, mattis laoreet lacus. [0m
  [2m                                                   Donec in quam tempus, eleifend erat sit amet, aliquet metus. 
[0m  [2m                                               Sed ullamcorper velit a est efficitur, et tempus ante rhoncus. 
[0m  [2m                                               Aliquam diam sapien, vulputate sit amet elit sit amet, commodo eleifend sapien. 
[0m  [2m                                               Donec consequat at nibh id mattis. Quisque vitae sagittis eros, convallis consectetur ante. 
[0m  [2m                                               Duis finibus suscipit mi sed consectetur. Nulla libero neque, sagittis vel nulla vel,
[0m[38;5;52m[48;5;225m-                                vestibulum sagittis mauris. Ut laoreet urna lectus. [0m
[38;5;52m[48;5;225m-                                    Sed lorem felis, condimentum eget vehicula non, sagittis sit amet diam. [0m
[38;5;52m[48;5;225m-                                    Vivamus ut sapien at erat imperdiet suscipit id a lectus.[0m
[38;5;22m[48;5;159m+                                   vestibulum sagittis mauris. Ut laoreet urna lectus. [0m
[38;5;22m[48;5;159m+                                   Sed lorem felis, condimentum eget vehicula non, sagittis sit amet diam. [0m
[38;5;22m[48;5;159m+                                   Vivamus ut sapien at erat imperdiet suscipit id a lectus.[0m
[38;5;22m[48;5;159m+                                   Another Line added.[0m

[2mat snap/path:10
[0m
---

[TestDiff/no_color/should_apply_highlights_on_single_line_diff - 1]

- Snapshot - 1
+ Received + 1

- abcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd
+ abcfabcfabcfabcfabcfabcfabcfabcfabcfabcfabcfabcfabcfabcfabcfabcfabcfabcfabcfabcf

at snap/path:10

---

[TestDiff/no_color/multiline_diff - 1]

- Snapshot - 5
+ Received + 6

@@ -1,12 +1,13 @@

  Proin justo libero, pellentesque sit amet scelerisque ut, sollicitudin non tortor. 
-       Sed velit elit, accumsan sed porttitor nec, elementum quis sapien. 
+       Sed velit elit, accumsan sed Ipsum nec, elementum quis sapien. 
        Phasellus mattis purus in dui pretium, eu euismod metus feugiat. 
-       Morbi turpis tellus, tincidunt mollis rutrum at, mattis laoreet lacus. 
+       Morbi turpis Lorem, tincidunt mollis rutrum at, mattis laoreet lacus. 
        Donec in quam tempus, eleifend erat sit amet, aliquet metus. 
        Sed ullamcorper velit a est efficitur, et tempus ante rhoncus. 
        Aliquam diam sapien, vulputate sit amet elit sit amet, commodo eleifend sapien. 
        Donec consequat at nibh id mattis. Quisque vitae sagittis eros, convallis consectetur ante. 
        Duis finibus suscipit mi sed consectetur. Nulla libero neque, sagittis vel nulla vel,
-        vestibulum sagittis mauris. Ut laoreet urna lectus. 
-        Sed lorem felis, condimentum eget vehicula non, sagittis sit amet diam. 
-        Vivamus ut sapien at erat imperdiet suscipit id a lectus.
+       vestibulum sagittis mauris. Ut laoreet urna lectus. 
+       Sed lorem felis, condimentum eget vehicula non, sagittis sit amet diam. 
+       Vivamus ut sapien at erat imperdiet suscipit id a lectus.
+       Another Line added.

at snap/path:20

---

[TestDiff/should_not_print_snapshot_line_if_not_provided - 1]

[38;5;52m[48;5;225m- Snapshot  - 2[0m
[38;5;22m[48;5;159m+ Received + 10[0m

there is a diff here

---

[TestDiff/should_print_newline_diffs/multiline - 1]

[38;5;52m[48;5;225m- Snapshot - 0[0m
[38;5;22m[48;5;159m+ Received + 3[0m

  [2msnapshot
[0m[38;5;22m[48;5;159m+ [0m
  [2m         with multiple lines
[0m[38;5;22m[48;5;159m+ [0m
[38;5;22m[48;5;159m+          diff[0m
  [2m                          
[0m
[2mat snap/path:10
[0m
---

[TestDiff/should_print_newline_diffs/multiline - 2]

[38;5;52m[48;5;225m- Snapshot - 3[0m
[38;5;22m[48;5;159m+ Received + 0[0m

  [2msnapshot
[0m[38;5;52m[48;5;225m- [0m
  [2m         with multiple lines
[0m[38;5;52m[48;5;225m- [0m
[38;5;52m[48;5;225m-          diff[0m
  [2m                          
[0m
[2mat snap/path:10
[0m
---

[TestDiff/should_print_newline_diffs/singleline - 1]

[38;5;52m[48;5;225m- Snapshot - 0[0m
[38;5;22m[48;5;159m+ Received + 1[0m

[48;5;225m[38;5;52m- [0m[48;5;225m[38;5;52msingle line snap[0m
[48;5;159m[38;5;22m+ [0m[48;5;159m[38;5;22msingle line snap[0m[48;5;23m[38;5;255m â†µ[0m

[2mat snap/path:10
[0m
---

[TestDiff/should_print_newline_diffs/singleline - 2]

[38;5;52m[48;5;225m- Snapshot - 0[0m
[38;5;22m[48;5;159m+ Received + 1[0m

[48;5;225m[38;5;52m- [0m[48;5;225m[38;5;52msingle line snap[0m
[48;5;159m[38;5;22m+ [0m[48;5;159m[38;5;22msingle line snap[0m[48;5;23m[38;5;255m â†µ[0m

[2mat snap/path:10
[0m
---

[TestDiff/should_print_newline_diffs/singleline - 3]

[38;5;52m[48;5;225m- Snapshot - 0[0m
[38;5;22m[48;5;159m+ Received + 1[0m

[48;5;225m[38;5;52m- [0m[48;5;225m[38;5;52msingle line snap[0m
[48;5;159m[38;5;22m+ [0m[48;5;159m[38;5;22msingle line snap[0m[48;5;23m[38;5;255mâ†µ[0m

[2mat snap/path:10
[0m
---

[TestDiff/should_print_newline_diffs/singleline - 4]

[38;5;52m[48;5;225m- Snapshot - 1[0m
[38;5;22m[48;5;159m+ Received + 0[0m

[48;5;225m[38;5;52m- [0m[48;5;225m[38;5;52msingle line snap[0m[48;5;127m[38;5;255mâ†µ[0m
[48;5;159m[38;5;22m+ [0m[48;5;159m[38;5;22msingle line snap[0m

[2mat snap/path:10
[0m
---
